src/main/java/org/apache/spark/shuffle/ucx/UcxNode.java

```diff
@@ -0,0 +1,238 @@
+ */
+public class UcxNode implements Closeable {
+  private final UcpContext context;
+  private final MemoryPool memoryPool;
+  private final UcpWorkerParams workerParams = new UcpWorkerParams().requestThreadSafety();
+  private UcpListener listener;
+  private final UcpWorker globalWorker;
+  private UcpEndpoint globalDriverEndpoint;
+
+  private boolean closed = false;
+
+  private static final Set<UcxWorkerWrapper> workerPool = ConcurrentHashMap.newKeySet();
+  private static final Map<BlockManagerId, ByteBuffer> workerAdresses = new ConcurrentHashMap<>();
+  private static final ConcurrentHashMap<UcpEndpoint, ByteBuffer> rpcConnections =
+    new ConcurrentHashMap<>();
+
+  private final ThreadLocal<UcxWorkerWrapper> worker;
+  private final Thread listenerProgressThread;
+
+  private static final Logger logger = LoggerFactory.getLogger(UcxNode.class);
+  private static final AtomicInteger numWorkers = new AtomicInteger(0);
+
+  public UcxNode(UcxShuffleConf conf, boolean isDriver) {
+    UcpParams params = new UcpParams().requestTagFeature()
+      .requestRmaFeature().requestWakeupFeature()
+      .setMtWorkersShared(true);
+    context = new UcpContext(params);
+    memoryPool = new MemoryPool(context, conf);
+    globalWorker = context.newWorker(workerParams);
+    InetSocketAddress driverAddress = new InetSocketAddress(conf.driverHost(), conf.driverPort());
+
+    if (isDriver) {
+      // 1. Start listener on a driver and accept RPC messages from executors with their
+      // worker addresses
+      UcpListenerParams listenerParams = new UcpListenerParams().setSockAddr(driverAddress);
+      listener = globalWorker.newListener(listenerParams);
+      logger.info("Started UcxNode on {}", driverAddress);
+    } else {
+      // 1. Executor: connect to driver using sockaddr
+      // and send it's worker address followed by BlockManagerID.
+      globalDriverEndpoint = globalWorker.newEndpoint(
+        new UcpEndpointParams().setSocketAddress(driverAddress).setPeerErrorHadnlingMode()
+      );
+      BlockManagerId blockManagerId = SparkEnv.get().blockManager().blockManagerId();
+      ByteBuffer workerAddresss = globalWorker.getAddress();
+
+      ByteBuffer metadataBuffer = Platform.allocateDirectBuffer(conf.metadataRPCBufferSize());
+      metadataBuffer.order(ByteOrder.nativeOrder());
+      metadataBuffer.putInt(workerAddresss.capacity());
+      metadataBuffer.put(workerAddresss);
+      try {
+        ObjectOutputStream oos = new ObjectOutputStream(
+          new ByteBufferOutputStream(metadataBuffer));
+        blockManagerId.writeExternal(oos);
+        oos.close();
+      } catch (IOException e) {
+        logger.error("Exception on serializing BlockManager: {}", e.getMessage());
+      }
+
+      // TODO: send using stream API when it would be available in jucx.
+      globalDriverEndpoint.sendTaggedNonBlocking(metadataBuffer, null);
```

Comment:
how to we make sure the reference to metadataBuffer is not released before the send is completed?

Reply:
but does ucx not guarantee, that if the address is not registered - to pin it ?

Reply:
in case of bcopy, the memory could be overriden